### PR TITLE
Adapt coq_makefile to handle coqpp-based macro files.

### PR DESCRIPTION
**Pull request by ppedrot**
This allows external plugins to use the coqpp tools in lieu of camlp5 macros. I haven't included a test yet because I am unsure how to do it properly, but the patch works on the adaptation of the test that can be found in the test-suite.

### CODE REVIEW

**Comment by ejgallego** [comment]
I'll let @gares handle the `_CoqProject` part.

File: tools/CoqMakefile.in [hunk 1]
```diff
@@ -602,6 +607,18 @@ $(ML4FILES:.ml4=.cmx): %.cmx: %.ml4
 	$(SHOW)'CAMLOPT -pp -c $(FOR_PACK) $<'
 	$(HIDE)$(CAMLOPTC) $(CAMLDEBUG) $(CAMLFLAGS) $(CAMLPKGS) $(PP) $(FOR_PACK) -impl $<
 
+$(MLGFILES:.mlg=.ml): %.ml: %.mlg
+	$(SHOW)'COQPP $<'
+	$(HIDE)$(COQPP) $<
+
+$(MLGFILES:.mlg=.cmo): %.cmo: %.ml
```
Comment (gares) [inline tools/CoqMakefile.in]: This part can be factored with the regular ml files if I'm not mistaken, eg `$(MLFILES:.ml=.cmo) $(MLGFILES:.mlg=.cmo): %.cmo: %.ml`

File: tools/CoqMakefile.in [hunk 2]
```diff
@@ -602,6 +607,18 @@ $(ML4FILES:.ml4=.cmx): %.cmx: %.ml4
...
+	$(SHOW)'CAMLC -c $<'
+	$(HIDE)$(CAMLC) $(CAMLDEBUG) $(CAMLFLAGS) $(CAMLPKGS) $<
+
+$(MLGFILES:.mlg=.cmx): %.cmx: %.ml
```
Comment (gares) [inline tools/CoqMakefile.in]: as above

**Review by gares — COMMENTED** [review]
looks pretty good to me

**Comment by gares** [comment]
I wanted to request for changes, I hit the wrong button, and now I can't request for changes I don't find other stuff to change ;-)

File: tools/CoqMakefile.in (hunk 1 above)
Comment (ppedrot) [reply tools/CoqMakefile.in]: This syntax doesn't seem to work.

Comment (gares) [reply tools/CoqMakefile.in]: maybe you have a ALLCMO files variable already containing that union...

**Review by gares — APPROVED** [review]

File: tools/CoqMakefile.in (hunk 1 above)
Comment (ppedrot) [reply tools/CoqMakefile.in]: I don't like expanding the set of global variables, I think I prefer the current hackish but local solution.

**Comment by ppedrot** [comment]
@gares You should probably assign that to yourself (and merge it if deemed OK).

**Comment by ejgallego** [comment]
@ppedrot, would you mind adapting the `.ml4` files in `test-suite` to this patch?

That is the only thing remaining for the test-suite to run in OCaml `trunk`.

**Comment by ppedrot** [comment]
Should I leave the legacy test for ml4 as well?

**Comment by ejgallego** [comment]
> Should I leave the legacy test for ml4 as well?

The thing is that it will make the test-suite fail in camlp5-less setups. So we would have to add some conditional machinery which is complex to do in make.

 IMHO as we are leading to ml4 deprecation we could just rely on the CI for testing ml4.

**Comment by ppedrot** [comment]
Fine, I'll remove it then.

**Comment by SkySkimmer** [comment]
It's a coq-makefile test so it's in shell.

Gaëtan Gilbert

**Comment by ppedrot** [comment]
@SkySkimmer Would still be hard as we need to generate the _CoqProject files on the fly, so a *usine à gaz* for no good reason in 6 months from now.

**Review by ejgallego — APPROVED** [review]
Looks good to me and ready for merging. I concur with @ppedrot that there is not point in keeping some complex code for ml4 testing as we may remove support for it in 8.11.

@ppedrot this makes `ml4` files officially deprecated then. Should we add indeed a deprecation warning to  the ml4 processor?

**Comment by gares** [comment]
> . Should we add indeed a deprecation warning to the ml4 processor?

yes

**Review by gares — APPROVED** [review]

**Comment by ppedrot** [comment]
@gares Can you merge, so that I start porting the plugins we track in the CI?

**Comment by ppedrot** [comment]
@gares ping.

**Pull request merged**